build-aux/snap/snapcraft.yaml

```diff
@@ -28,172 +29,294 @@ license: GPL-3.0
+    source: .
+    build-snaps:
+      - go/1.18/stable
+    after:
+      - apparmor
+    # these packages are needed to call mkversion.sh in override-pull, all other
+    # dependencies are installed using apt-get build-dep
+    build-packages:
+      - git
+      - dpkg-dev
+      - autoconf
+      - autoconf-archive
+      - automake
+      - xfslibs-dev
+      - libudev-dev
+      - libcap-dev
+      - libseccomp-dev
+      - pkg-config
+    override-pull: |
+      craftctl default
+      # set version, this needs dpkg-parsechangelog (from dpkg-dev) and git
+      craftctl set version="$(./mkversion.sh --output-only)"
+      ./get-deps.sh --skip-unused-check
+    override-build: |
+      ./mkversion.sh
+      cd "${CRAFT_PART_BUILD}/cmd"
+
+      autoreconf -fvi
+      ./configure --prefix=/usr --libexec=/usr/lib/snapd --without-unit-tests --enable-static-libapparmor --disable-host-binaries
+      make -j"${CRAFT_PARALLEL_BUILD_COUNT}"
+      make -j1 DESTDIR="${CRAFT_PART_INSTALL}" install
+
+      cd "${CRAFT_PART_BUILD}"
+
+      # Work-around for snapcraft not working with git worktrees
+      # We detect it with .git being a file rather than a directory
+      if [ -f .git ]; then
+        EXTRA_GO_FLAGS="-buildvcs=false"
+      fi
+      CMDS=(bin/snap
+            lib/snapd/snapd
+            lib/snapd/snap-exec
+            lib/snapd/snap-failure
+            lib/snapd/snap-fde-keymgr
+            lib/snapd/snap-preseed
+            lib/snapd/snap-recovery-chooser
+            lib/snapd/snap-repair
+            lib/snapd/snap-seccomp
+            lib/snapd/snap-update-ns
+            lib/snapd/snapctl
+            lib/snapd/snapd-apparmor)
+      for cmd in "${CMDS[@]}"; do
+        case "${cmd}" in
+           # FIXME: some binaries need to be run confined in apps. But
+           # instead we should allow apps to access dynamic linkers
+           # and libraries from snapd.
+           lib/snapd/snap-exec|lib/snapd/snapctl)
+             export CGO_ENABLED=0
+             GO_LD_FLAGS=()
+             ;;
+           lib/snapd/snap-update-ns)
+             export CGO_ENABLED=1
+             GO_LD_FLAGS=(--ldflags '-extldflags "-static"')
+             ;;
+           *)
+             export CGO_ENABLED=1
+             GO_LD_FLAGS=()
+             ;;
+        esac
+
+        case "${cmd}" in
+          bin/snap)
+            TAGS=(nomanagers)
+            case "$(craftctl get version)" in
+              1337.*)
+                TAGS+=(withtestkeys faultinject)
+                ;;
+            esac
+            ;;
+          *)
+            TAGS=()
+            case "$(craftctl get version)" in
+              1337.*)
+                TAGS+=(withtestkeys withbootassetstesting faultinject)
+                ;;
+            esac
+            ;;
+        esac
+
+        go build -mod=vendor -tags "${TAGS[*]}" "${GO_LD_FLAGS[@]}" ${EXTRA_GO_FLAGS-} -o "${CRAFT_PART_INSTALL}/usr/${cmd}" "github.com/snapcore/snapd/cmd/$(basename ${cmd})"
```

Comment:
just to be sure, I'd add a snippet checking if snap-update-ns, snap-exec and snapctl are indeed statically linked. IIRC -static only enabled static linking but not required it. Again, we do this in !Ubuntu/Debian.

Reply:
I have added a check.